tool/infocert.c

```diff
@@ -0,0 +1,108 @@
+ *
+ * Copyright 2018 Xaptum, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+#include "infocert.h"
+#include <xtt/crypto_types.h>
+#include <xtt/certificates.h>
+#include <xtt/util/asn1.h>
+#include <xtt/util/root.h>
+#include <xtt/util/util_errors.h>
+#include <xtt/util/file_io.h>
+#include <string.h>
+#include <stdio.h>
+
+struct xtt_server_certificate_raw_type;
+
+static
+void printf_hex(unsigned char* raw, size_t len)
+{
+    for (size_t i = 0; i < len; i++) {
+        printf("%02x", raw[i]);
+    }
+    printf("\n");
+}
+
+static
+void printf_char(unsigned char* raw, size_t len)
+{
+    for (size_t i = 0; i < len; i++) {
+        printf("%c", raw[i]);
+    }
+    printf("\n");
+}
+
+static
+void server_cert_out(struct xtt_server_certificate_raw_type* certificate)
+{
+    unsigned char* server_id = xtt_server_certificate_access_id(certificate);
+    printf("Server ID: ");
+    printf_hex(server_id, sizeof(xtt_identity_type));
+
+    unsigned char* expiry = xtt_server_certificate_access_expiry(certificate);
+    printf("Expiry: ");
+    printf_char(expiry, sizeof(xtt_certificate_expiry));
+
+    unsigned char* root_id = xtt_server_certificate_access_rootid(certificate);
+    printf("Root ID: ");
+    printf_hex(root_id, sizeof(xtt_certificate_root_id));
+
+    unsigned char* public_key = xtt_server_certificate_access_pubkey(certificate);
+    printf("Public Key: ");
+    printf_hex(public_key, sizeof(xtt_ecdsap256_pub_key));
+
+    return;
+}
+
+static
+void root_cert_out(xtt_root_certificate* root_certificate)
+{
+    xtt_certificate_root_id root_id;
+    xtt_ecdsap256_pub_key public_key;
+    xtt_deserialize_root_certificate(&public_key, &root_id, root_certificate);
+    printf("Root ID: ");
+    printf_hex(root_id.data, sizeof(xtt_certificate_root_id));
+
+    printf("Public key: ");
+    printf_hex(public_key.data, sizeof(xtt_ecdsap256_pub_key));
+    return;
+}
+
+
+int info_cert(const char* filename, enum infocert_type type)
+{
+    int read_ret = 0;
+
+    if (type == infocert_type_server) {
+        unsigned char certificate[XTT_SERVER_CERTIFICATE_ECDSAP256_LENGTH];
+        read_ret = xtt_read_from_file(filename, certificate, XTT_SERVER_CERTIFICATE_ECDSAP256_LENGTH);
+        if(read_ret != XTT_SERVER_CERTIFICATE_ECDSAP256_LENGTH) {
```

Comment:
This check will fail for a file that is too short, but not for a file that is too long.

For instance, the tool will happily process a server cert as though it were a root cert (since a server cert is longer):

```
build/tool$ ./xtt infocert --rootcert ~/git/xmb/xmb/conf/dev/xtt_certificate.bin
Root ID: 58415054554d54455354534552563031
Public key: 323130303132333158415054554d54455354524f4f5431300445013c60ee505c0236eda88440a8ad08e4cc843d37bc27c576a13626fe9031be6c959714591045bf
ok
```
But `xtt_certificate.bin` is a server cert, not a root cert.

Instead, we want to fail if the file is too short or if it is too long.

Perhaps `xtt_read_from_file` should return an error if the file is longer than the provided buffer.  You'll need to audit all the uses of `xtt_read_from_file` to make sure this change is safe.  If it is, it can go in a separate commit within the PR.